### ISSUE TITLE
Release 1.7.x and 2.0.0

### ISSUE THREAD

**Issue by ericpre** [hyperspy/hyperspy#2996]
## [v2.0.0](https://github.com/hyperspy/hyperspy/milestone/31)
As the IO have been split - see https://github.com/hyperspy/hyperspy/pull/2972 and https://github.com/hyperspy/rosettasciio - we should get 2.0 released ASAP.

As this is major release, it is a good opportunity to change some API
- [x] Split spectroscopy (EDS and EELS) #3202
- [x] Split holography #3201 
- [x] Add intersphinx links to HoloSpy/EleXSpy ones available (e.g. for mentions introduced in #3216)
- [x] Clear all deprecations
  - [x] hyperspy/hyperspy#2998
  - [x] hyperspy/hyperspy#2999
  - [x] hyperspy/hyperspy#3000
  - [x] hyperspy/hyperspy#3007
  - [x] hyperspy/hyperspy#3014
  - [x] hyperspy/hyperspy#3017
  - [x] hyperspy/hyperspy#3037 of various functions and arguments
  - [x] hyperspy/hyperspy#3063
  - [x] Check for leftovers of the deprecation (especially in the documentation)
    - [x] Correct documentation and examples on `optimizer` and `loss_function` parameters in user guide (model-fitting page) and docstring of `m.fit()`
    - [x] fix marker naming in docstring examples (done in hyperspy/hyperspy#3063)
- [ ] go through some API changes worth doing
    - [x] #3223
    - [ ] #3170 
    - [ ]  for example https://github.com/hyperspy/hyperspy/issues/1227
    - [x] #3197 rename `crop_image` to `crop_signal2D` to be consistent with `crop_signal1D`?
    - [ ] reorganize metadata: https://github.com/hyperspy/hyperspy/issues/2095
    - [ ] review usage of `mask` versus `navigation_mask`, `signal_mask` and if consistency can be improved - see for example https://github.com/hyperspy/hyperspy/issues/3053
- [x] Clarify what is public/private, and improve API documentation?
      - See for example: https://github.com/hyperspy/hyperspy/pull/3014#discussion_r973575789
- [ ] Improve Plotting of Markers/ peakfinding:
    - [x] hyperspy/hyperspy#3148 
    - [x] hyperspy/hyperspy#3142
    - [ ] #3196 
    - [x] hyperspy/hyperspy#3175 
    - [x] #3140 
    - [ ] #3199 
- [x] Dependency Management:
    - [x] hyperspy/hyperspy#3145 
- [ ] Improve Axes Handling 
    - [ ] hyperspy/hyperspy#2756 
    - [ ] hyperspy/hyperspy#3031  (this could be dropped/ refactored)
    - [ ] hyperspy/hyperspy#3055 (this could be dropped/ refactored)
   - [ ] #3286
- [x] Fix Map parallel and lazy bugs
    - [x] #3198 (closes hyperspy/hyperspy#3152  and hyperspy/hyperspy#2997)
    - [x] hyperspy/hyperspy#2971
    - [x] review `rechunk` argument of lazy functions. hyperspy/hyperspy#3166
   ~~ - [ ] hyperspy/rosettasciio#140 ~~ Non-Blocking
   ~~ - [ ] hyperspy/rosettasciio#141 ~~ Non-Blocking

- [ ] Good things worth doing:
    - [x] hyperspy/hyperspy#3140
    - [x] hyperspy/hyperspy#3082 
    - [ ] hyperspy/hyperspy#3030
    - [x] hyperspy/hyperspy#3099
    - [x] #3174 

- [ ] Check with libraries relying on hyperspy that they are happy with the 2.0 release
    - [x] Release Candidate published.
- [x] Review release notes, particularly consistency in the API removal, etc.

Anything else, which would be worth considering, please edit or add a comments in this issue!

## [v1.7.4](https://github.com/hyperspy/hyperspy/milestone/50) (released)
There are several bug fixes in the `RELEASE_next_patch` branch and it would be good to make a quick patch release
- [x] https://github.com/hyperspy/hyperspy/pull/3103
- [x] https://github.com/hyperspy/hyperspy/pull/3102
- [x] https://github.com/hyperspy/hyperspy/pull/3062
- [x] https://github.com/hyperspy/hyperspy/pull/3109
- [x] https://github.com/hyperspy/hyperspy/pull/3108

It would be good to merge these PRs:
- [x] https://github.com/hyperspy/hyperspy/pull/3073

There are a number of PRs merged into `RELEASE_next_major` that would make sense to add in a patch release:
- [x] hyperspy/hyperspy#3084

**Comment by hakonanes** [comment]
> Check with libraries relying on hyperspy that they are happy with the 2.0 release

We haven't tested kikuchipy with HyperSpy 2.0, but I will report back here once we have. Is there anything special we should be aware of?

**Comment by CSSFrancis** [comment]
@ericpre There are a couple of libraries that would benefit from hyperspy/hyperspy#2876. I'd like for the `find_peaks` method to just return a vector signal. As written there shouldn't be many conflicts down stream but it is probably best to make that change during the major release. 

Pyxem is trying to release a 1.0.0 version as well so my hope is that will closely follow the 2.0 release here.

As far as other things to consider. I think in hyperspy/hyperspy#821 there was also some talk about splitting off the EDS and EELS signals and models. This might be worth a discussion as well. I feel it might be hurting our chances of being used as a general high dimensional image viewer.

**Comment by CSSFrancis** [comment]
It might also be a good idea to formalize how Hyperspy handles parallelization.   At the very least I think that this involves more documentation about how to use `dask` as well as a discussion on the different schedulers. Currently, every function calling `map` is parallel by default because `dask` is parallel by default even if parallel=False. (Some call this a bug, some call this a feature!)

I think this is a good thing as it gives the user the ability to set up `dask` however works for them (or not set it up and have it just work by default). 

The only other place where multiprocessing comes in is with the SamFIRE utils.  I've given those a small look, and think that switching those to `dask` might also be possible. That would simplify the code a fair amount and really streamline all more advanced operations through `dask` rather than relying on custom built parallelization code. 

 I'll create an issue for this and link it here. (https://github.com/hyperspy/hyperspy/issues/2997)

**Comment by CSSFrancis** [comment]
> It might also be a good idea to formalize how Hyperspy handles parallelization. At the very least I think that this involves more documentation about how to use `dask` as well as a discussion on the different schedulers. Currently, every function calling `map` is parallel by default because `dask` is parallel by default even if parallel=False. (Some call this a bug, some call this a feature!)
> 
> I think this is a good thing as it gives the user the ability to set up `dask` however works for them (or not set it up and have it just work by default).
> 
> The only other place where multiprocessing comes in is with the SamFIRE utils. I've given those a small look, and think that switching those to `dask` might also be possible. That would simplify the code a fair amount and really streamline all more advanced operations through `dask` rather than relying on custom built parallelization code.

I'll create an issue for this and link it here.

**Comment by ericpre** [comment]
> As far as other things to consider. I think in https://github.com/hyperspy/hyperspy/issues/821 there was also some talk about splitting off the EDS and EELS signals and models. This might be worth a discussion as well. I feel it might be hurting our chances of being used as a general high dimensional image viewer.

This is definitely something that will happen but I don't know if this would be possible in a reasonable timeframe for 2.0.

> I'll create an issue for this and link it here. (https://github.com/hyperspy/hyperspy/issues/2997)

Thanks @CSSFrancis for this suggestion, indeed, it would be to improve it.

**Comment by ericpre** [comment]
> > Check with libraries relying on hyperspy that they are happy with the 2.0 release
> 
> We haven't tested kikuchipy with HyperSpy 2.0, but I will report back here once we have. Is there anything special we should be aware of?

@hakonanes, it may be better to check after the deprecation and api changes have been completed.

**Comment by francisco-dlp** [comment]
Major releases force people to update their codes and scripts, so we should minimize their number. Therefore, I think that it would be good to split the TEM features into a different package for the 2.0 release. It shouldn't be too hard. I think that splitting them into a single package would be good. What do others think? The biggest difficulty is to find a name for such a package.

**Comment by CSSFrancis** [comment]
> Major releases force people to update their codes and scripts, so we should minimize their number. Therefore, I think that it would be good to split the TEM features into a different package for the 2.0 release. It shouldn't be too hard. I think that splitting them into a single package would be good. What do others think? The biggest difficulty is to find a name for such a package.

I'm in support of it. In terms of names I like the direction the Lumispy took in keeping the -spy as that suggests the connection between the package and hyperspy.

So maybe electrospy, elspy or espy.

Spectrospy? Spectspy

**Comment by hakonanes** [comment]
Realize I might have less to say on the matter since I’m not a core developer of HyperSpy, but keeping the HyperSpy name for the spectroscopy package and calling the core package HyperSpy-core/base is also an option.

**Comment by ericpre** [comment]
Splitting all domain-specific signals would be great. A few comments:

- I don't think that it would be possible to complete https://github.com/hyperspy/hyperspy/pull/2836 in a reasonable timeframe, so this PR may need to be redone after the split.
- `HologramImage` is in principle not limited to electron microscopy, so it may be better that it goes to a separate package `hyperspy-holo`.
- A good word to describe both EDS and EELS without mentioning them explicitly is "analytical electron microscopy".

**Comment by jlaehne** [comment]
> Splitting all domain-specific signals would be great. A few comments:
Yes, I also wanted to bring up these additional splits as well. I think it makes totally sense to do them now (even though I am not happy delaying the 2.0 release as I want luminescence signals from rosettascio to be available to lumispy through v2.0)

>     * `HologramImage` is in principle not limited to electron microscopy, so it may be better that it goes to a separate package `hyperspy-holo`.
Yes, I think holography is something completely different than EELS and EDX.

>     * A good word to describe both EDS and EELS without mentioning them explicitly is "analytical electron microscopy".
However, for me also CL and EBSD are analytical electron microscopy, but already contained in separate extensions.

> So maybe electrospy, elspy or espy.
I like ElSpy or ElectronSpy or even eSpy

**Comment by ericpre** [comment]
I think this is time to decide if we split into 2 or 3 libraries:

| 3 libraries   (vote :rocket: )   |  2 libraries (vote :tada:)            |
| -------------------| -----------------|
| hyperspy       | hyperspy    |
| spectroscopy (EDS and EELS)         | spectroscopy (EDS and EELS) and holography  |
| holography         |     |


Going for the 3 libraries option has the advantage of having the scope of the package well defined, while having holography in the same library as EELS and EDS sounds like an after thought. There will be some overhead with maintaining one more library, but we are well experienced with having a lot of the testing and packaging automated, so it will not be much. And we never know, as some point, some people may be interested to develop the holography library and this could be advertised as such: "maintenance only and looking for developer".

This also has the advantage of putting spectroscopy techniques into a same library and this would good to be able to add XRF signal to this spectroscopy package - if we go for the 2 libraries option, it may end up being focused on electron based technique and exclude XRF, for example. This has been discussed several times already and there are many similarities between EDS and XRF.

As silver lining, we can use the `-spy` pattern (see https://github.com/hyperspy/hyperspy/issues/2996#issuecomment-1228504094), such as `holospy` and `spectrospy` are available on pypi! :)

**Comment by jlaehne** [comment]
I would like to reorganize the metadata for SEM and TEM related signals as proposed in: https://github.com/hyperspy/hyperspy/issues/2095

The major version and splitting would be the right moment to address this issue!

(I have added it to the checklist above)

**Comment by ericpre** [comment]
Because of https://github.com/hyperspy/hyperspy/pull/3016, we should release 1.7.2 ASAP. I updated the [summary above](https://github.com/hyperspy/hyperspy/issues/2996#issue-1350647756).

**Comment by jlaehne** [comment]
> Because of hyperspy/hyperspy#3016, we should release 1.7.2 ASAP. I updated the [summary above](https://github.com/hyperspy/hyperspy/issues/2996#issue-1350647756).

See hyperspy/hyperspy#3018

**Comment by ericpre** [comment]
To track the failure of libraries depending on hyperspy: https://github.com/hyperspy/hyperspy-extensions-list/actions?query=branch%3Aadd_release_next_major

**Comment by CSSFrancis** [comment]
Do we have an idea of when we want to shoot for a 2.0.0 release?  We would like to release a 1.0.0 version of pyxem at the same time or shortly after.  I can devote a fair bit of time to this over the next couple of months if necessary to get this moving.  I think that at the very latest we should make sure that the 2.0.0 release occurs before M&M in July.

**Comment by jlaehne** [comment]
I think the main issue is that we still need to split the EELS/EDS and Holography libraries into extension packages to do all the reorganization in a single major release. See discussion above.

Also hyperspy/hyperspy#2095 (metadata conventions) would need to be addressed before a major release, there should not be much to do on the HyperSpy side, but it would entail a bit of work on the rosettasciio end.

In any case, I do support not letting that trail too long, though my capacities in the coming months are a bit limited.

(I edited the todo-list above)

**Comment by ericpre** [comment]
I would suggest to make a patch release soon, because the current hyperspy released version is broken for lazy generating random number generator with the recent dask release. There should be easy workaround around and it shouldn't concern that much user code but it would be to make a patch release.
I have updated the first comment in this issue to track the PRs that need/could be completed.

**Comment by jlaehne** [comment]
> I would suggest to make a patch release soon, because the current hyperspy released version is broken for lazy generating random number generator with the recent dask release. There should be easy workaround around and it shouldn't concern that much user code but it would be to make a patch release. I have updated the first comment in this issue to track the PRs that need/could be completed.

I added a list of PRs merged into RnMajor in the past months that would make sense to add to a patch release already.

**Comment by ericpre** [comment]
After checking the list above, I suspect that the major and patch branch have diverge significantly, that it may be a pain to correctly backport these changes... and I am not sure that they are worth the effort. From the list above I would backport only hyperspy/hyperspy#3084. @jlaehne, what do you think?

**Comment by jlaehne** [comment]
Yes, that is probably the easiest and most important one.

We should then probably focus on getting the 2.0 release out soon as well.

**Comment by CSSFrancis** [comment]
> We should then probably focus on getting the 2.0 release out soon as well.

Is there anything that I can do to help this along?  I would like to get this figured out sooner rather than later. I can work on cleaning up my code additions but there are some points that I would like some additional input before finishing.

I don't know if I am good enough git wizard to split out the eels and eds functions but I can help organize and set up the extensions after the fact.  There are a couple of meetings/conferences/hyperspy workshops coming up and it would be good to get some of these changes in and packaged before then.

**Comment by jlaehne** [comment]
> Is there anything that I can do to help this along? I would like to get this figured out sooner rather than later. I can work on cleaning up my code additions but there are some points that I would like some additional input before finishing.

Well, if you feel comfortable with any of the proposed API changes mentioned above, you could help get that of the table. Also, it would be good to release RosettaSciIO soon: I created a tracking issue https://github.com/hyperspy/rosettasciio/issues/99 and there are still a number of improvements that could be done there.
 
> I don't know if I am good enough git wizard to split out the eels and eds functions but I can help organize and set up the extensions after the fact. There are a couple of meetings/conferences/hyperspy workshops coming up and it would be good to get some of these changes in and packaged before then.

Once RosettaSciIO is released, we need to decide whether we keep the full split for v2.0 or whether we actually wait with the split out of eels, eds and holography for a v3.0. It would be nice to make the full split now, but at the pace that we are at currently that would delay v2.0 quite a bit. The IO split was more work than anticipated, but then we also got a lot of things much cleaner on both the HyperSpy and RosettaSciIO side along the way.

**Comment by CSSFrancis** [comment]
At this point I think we should start thinking about releasing a 2.0.0 without splitting out the EDS, EELS and Holography.  The last major release was over a year ago at this point and there are a fair number of improvements since then.

I've edit the post above to include what I think v2.0.0 should have included.

Looking at these I realize that I might be part of the problem (rather than part of the solution) I will try to make sure all of my open PR's are passing and maybe we can figure out how to best review them as some of them are a bit more involved. Honestly some of them could very easily be added in a minor release as well.

I propose that 3.0.0 includes the EDS, EELS and Holography split and a more formalized metadata structure.  We should start looking for someone with interest to take over as a lead developer for that package as well. More consistent and smaller releases would probably do some good looking forward. And all of the splitting should help with that quite a bit.

**Comment by magnunor** [comment]
I made a pull request to address the ` - review rechunk argument of lazy functions.`:  https://github.com/hyperspy/hyperspy/pull/3166

**Comment by CSSFrancis** [comment]
I might try to start splitting the EELS/EDS as well as the holography unless someone else wants to give it a shot or someone says that we should wait.  It shouldn't be terribly difficult to accomplish (hopefully) and then that will bring us a fair bit closer to a 2.0.0 release.

**Comment by CSSFrancis** [comment]
> I might try to start splitting the EELS/EDS as well as the holography unless someone else wants to give it a shot or someone says that we should wait. It shouldn't be terribly difficult to accomplish (hopefully) and then that will bring us a fair bit closer to a 2.0.0 release.

Most of the errors in splitting this result from the pretty heavy use of the EDS_SEM signal https://github.com/hyperspy/hyperspy/blob/70ca0762a332a7239a4c1eb75001062ab6c7495d/hyperspy/datasets/example_signals.py#L2-L3 for testing the `model` functions.  The `create_model` function for the EDSSignal is a bit difficult to replicate.  An easy solution would be to depend on the EELS/EDS package for testing and then we could maybe slowly replace some of those tests.

**Comment by jlaehne** [comment]
> I might try to start splitting the EELS/EDS as well as the holography unless someone else wants to give it a shot or someone says that we should wait. It shouldn't be terribly difficult to accomplish (hopefully) and then that will bring us a fair bit closer to a 2.0.0 release.

Francisco also wanted to work on the EELS/EDS splitting when he finds the time - so maybe you could first start out with the holography part?

**Comment by CSSFrancis** [comment]
@jlaehne that's not a bad idea.  Splitting the Holography should be rather easy. 

My original goal of releasing 2.0.0 before M&M next week isn't looking terribly promising but we could maybe see about a release candidate which includes things like #3148.

**Comment by CSSFrancis** [comment]
So this actually seems like most things are pretty much done...

The big things left to do are:

- [ ] Split the EELS and EDS
- [ ] Split the Holography
- [ ] Clean up the `map` function
- [ ] Merge a bunch of outstanding code :) 

Beyond that we probably should try to ship something as there are lots of big things that are just waiting on this release.

@francisco-dlp or @ericpre were you thinking about tackling the splitting of the EELS and EDS code? 

I was going to clean up the `map` function today which (hopefully) shouldn't be too terribly difficult. 

Otherwise I will make another attempt at splitting things once I finish up with `map` function.

**Comment by ericpre** [comment]
I have put a due date for the [2.0 milestone](https://github.com/hyperspy/hyperspy/milestone/31) to mid-September. I have pick that date, because in August, there are holidays or conferences and after that we may have a window where we can focus on getting the release out of the door, without being distracted by other work activity - which for some of us will mostly start mid-/late-September.

Regarding testing of library dependency, it should be very easy/straightforward to install the development branch of hyperspy for testing purposes because no compilation is required anymore (this has moved to rossettasciio, which is now available on pypi and conda-forge).

We could do a [release candidate](https://devguide.python.org/developer-workflow/development-cycle/#release-candidate-rc), to catch potential bugs - but the final release should not include new features not already included in the RC.

**Comment by CSSFrancis** [comment]
> I have put a due date for the [2.0 milestone](https://github.com/hyperspy/hyperspy/milestone/31) to mid-September. I have pick that date, because in August, there are holidays or conferences and after that we may have a window where we can focus on getting the release out of the door, without being distracted by other work activity - which for some of us will mostly start mid-/late-September.

That sounds good to me.  My goal is to get hyperspy 2.0.0 and pyxem 1.0.0 wrapped up within the next month or 2 and then focus on writing for the next couple of months so I can graduate :) 

> Regarding testing of library dependency, it should be very easy/straightforward to install the development branch of hyperspy for testing purposes because no compilation is required anymore (this has moved to rossettasciio, which is now available on pypi and conda-forge).

Yea I've already done this a little bit with pyxem.  The release candidate would be nice for the CI but also you can just depend on the branch as well so it doesn't really matter

> We could do a [release candidate](https://devguide.python.org/developer-workflow/development-cycle/#release-candidate-rc), to catch potential bugs - but the final release should not include new features not already included in the RC.

Yea it might be a nice thing to consider going forward but I think most of the downstream packages already are tested with the integration tests so the release candidate is kind of redundant.  It might be nice for people testing workflows etc but I/m not sure how many people are even going to use it.

**Comment by hakonanes** [comment]
> Yea [an RC candidate] might be a nice thing to consider going forward but I think most of the downstream packages already are tested with the integration tests so the release candidate is kind of redundant.

Some breaking changes (marker collection e.g.) slated for the 2.0 release are not available in the dev branch but in unmerged PRs. It might be best therefore to make a release candidate with some more time to accomodate these changes for downstream packages.

**Comment by hakonanes** [comment]
From the activity here it seems like you're doing a great job towards the 2.0 release! Is it possible that HyperSpy 2.0 will be released on September 14 (as the "2.0 Split" milestone suggests)?

I'm asking so that I can plan a bit ahead. kikuchipy is still not compatible with the major release branch of HyperSpy. People downloading kikuchipy after HyperSpy 2.0 is available will thus find it unusable.

**Comment by jlaehne** [comment]
HoloSpy is close to release, but not yet released, while exSpy/EleXSpy will still need some work. So there will be some further delay in 2.0 release and September 14 is definitely too ambitious by now.

**Comment by CSSFrancis** [comment]
Unfortunately the sticking point seems to be #3148 😅  I think that #3202 is actually fairly close to being done once we can merge #3148.

**Comment by CSSFrancis** [comment]
@ericpre Do we have a good idea of what is left to be done here:

The minimum list is:
- [x] #3148 Needs a review of documentation and then merged
- [ ] #3202 Split the EELS EDX from hyperspy 
- [ ] #3196 Potentially clean up and make more consistent with find_peaks (for 2D signals)
- [x] #3233 

~Other low hanging fruit/ things I want in: (But aren't breaking changes)~

For 2.1.0 release:

- [ ] #3199 
- [ ] #2756

**Comment by ericpre** [comment]
We need to add #3233, as this is an API break, but it shouldn't be very difficult. Better done before the exSpy split, as this touch on that part of the code.

I think #3153 is not required and we need to get 2.0 out!

**Comment by CSSFrancis** [comment]
> I think #3153 is not required and we need to get 2.0 out!

Agreed :) I guess we go for the minimum and then some of the new features can come in 2.1.0 a little later.

**Comment by CSSFrancis** [comment]
This seems like we are getting closer to a 2.0.0 release!  What is the best way to handle releases downstream? Once #3202  and maybe #3197  and #3196  are merged we can always make a release candidate for testing downstream.

My plan was to:
- Create a final version of pyxem (0.16.0) that pins hyperspy to 1.x.x 
- Release a version of pyxem (0.17.0) with a minimum of hyperspy 2.x.x

What are other peoples thoughts on the best way to do this?

**Comment by ericpre** [comment]
Yes, this sounds like a good plan:
1. release hyperspy 2.0
2. updates extensions and release holospy, exspy, rosettasciio
3. once we are happy with everything, release hyperspy 2.0
4. make a bundle

I think that it may be worth making a bundle once pyxem 0.16 is released.

**Comment by magnunor** [comment]
I'll try to get in a release of Atomap as soon as possible as well, the marker collections are going to greatly improve the plotting of images with lots of atoms.

(In addition I just drastically improved the performance of fitting with 2D Gaussian in Atomap, for 2k x 2x images it will probably be 10 times as fast.)

**Comment by francisco-dlp** [comment]
With the new release there is a risk that the new version will generate files that cannot be read with version < 2.0. Also, we may be breaking backward file compatibility without realizing it because backward file compatibility is not thoroughly tested for all objects that HyperSpy can save into a hspy file. 

Therefore, I think that, before releasing we should:

- [ ] Document the API changes that lead to changes in the hspy metadata specification (in RosettaSciIO (!))
- [ ] Make sure that we can properly load old files, including EELS models, with the new version
- [ ] Bump the metadata version number (this is now in RosettaSciIO (!))
- [ ] Implement reading of the new metadata spec in a 1.7 patch release, for those who will prefer waiting before switching to 2.0

On this topic, I think that the primitive metadata specification management system in HyperSpy was manageable while HyperSpy was a monolithic package. Since we implemented the extensions, it is not longer the case. Indeed, any changes made to objects declared by extensions that are recorded in an hspy file constitute a metadata specification change. Currently we have no way of tracking those changes. In hyperspy, we make sure that file loading is backward compatible for HyperSpy objects. However, API changes in extensions may break backward file compatibility. The situation will only get worst with 2.0.

For more on this see https://github.com/LumiSpy/lumispy/issues/53#issuecomment-814772350 and the very interesting discussion in https://github.com/hyperspy/rosettasciio/issues/89 . I think that with 2.0, splitting the metadata specification into its own separate repository becomes urgent.

**Comment by francisco-dlp** [comment]
What would you think about adding a few more API changes to the release:

- [ ] Rename `Model.__call__` to `Model._get_current_data`  and `Signal.__call__` to `Model._get_current_data`. I think that the current syntax, implemented in the early stages of the development of EELSLab, is very unpythonic, hence unfriendly.
- [ ] Make the returns of `plot_images` and `plot_spectra` consistent. Currently the former returns an array of axes and the latter a list of axes.

**Comment by ericpre** [comment]
I would expect that for _some_ files generated with hyperspy 1.7 it will not be possible read it with hyperspy 1.3 (for example), if it contains components which has been implemented b etween version 1.3 and 1.7. However, when removing deprecated component, we did generate files with various version of hyperspy containing old components, etc. to make sure that future version of hyperspy will read these old files. 

> * [ ]  Make sure that we can properly load old files, including EELS models, with the new version

It would be good to add some hyperspy 1.7 file with eelsmodel as part of https://github.com/hyperspy/hyperspy/pull/3202, particularly if the API of EELSmodel changes because of the convolution implementation.

> * [ ]  Document the API changes that lead to changes in the hspy metadata specification (in RosettaSciIO (!))
> * [ ]  Bump the metadata version number (this is now in RosettaSciIO (!))

Is there anything that is missing from https://hyperspy.org/rosettasciio/supported_formats/hspy.html#changelog?

> * [ ]  Implement reading of the new metadata spec in a 1.7 patch release, for those who will prefer waiting before switching to 2.0

Can you elaborate?

**Comment by ericpre** [comment]
> What would you think about adding a few more API changes to the release:
> 
> * [ ]  Rename `Model.__call__` to `Model._get_current_data`  and `Signal.__call__` to `Model._get_current_data`. I think that the current syntax, implemented in the early stages of the development of EELSLab, is very unpythonic, hence unfriendly.
> * [ ]  Make the returns of `plot_images` and `plot_spectra` consistent. Currently the former returns an array of axes and the latter a list of axes.

These are a sensible changes to do, but at some point, we need to stop adding more API changes to be able to do the release. there are already API changes that we pushed to 3.0.

**Comment by CSSFrancis** [comment]
> These are a sensible changes to do, but at some point, we need to stop adding more API changes to be able to do the release. there are already API changes that we pushed to 3.0.

😅 Maybe I'm a little too ambitious but it would be nice to get this done by the end of this weekend or next week at the latest.  At some point @ericpre is right, we can make changes forever and never go anywhere. But there are lots of features that it would be nice to have in the bundle/released so I can just point people there rather than trying to convince people to download from GitHub.

I think the file loading consistency with `rosettasciio` is important.  It would be nice to be able to load future file versions but that sounds difficult.  I'd rather just include a warning that in order to load this file they might need to install a new version. That shouldn't hold up the release though.  That's the nice thing about splitting is that we can (hopefully) release more smaller version updates.

**Comment by francisco-dlp** [comment]
I will start working on the `__call__` methods.

**Comment by ericpre** [comment]
I think that we only have https://github.com/hyperspy/hyperspy/pull/3253 left before doing a hyperspy 2.0 release candidate. After that, we can release/update the extensions.

**Comment by francisco-dlp** [comment]
When splitting the IO we solved the issue of what to do with `DictionaryTreeBrowser`, that both `RosettaSciIO` and `HyperSpy` use, by using `python-box` in RosettaSciIO instead ([see the code here](https://github.com/hyperspy/rosettasciio/blob/697cea42e4ec8510c57aa0bf1996bd8bf0a65d7c/rsciio/utils/tools.py#L376)). This has the advantage of relying on a well maintained, standard package for this feature with plenty of advanges (including performance) over our overengineered `DictionaryTreeBrowser`. I suggest to do the same in HyperSpy for the 2.0 release. I can work on a PR if it sounds good.

**Comment by CSSFrancis** [comment]
@francisco-dlp I'd say if it's a quick change go for it. We should be able to make the change without breaking the API though if we support things like searching the Python box. Correct?

**Comment by francisco-dlp** [comment]
While it is possible to mimic the `DictionaryTreeBrowser` syntax while using `box` as you mention, I think that, in this case, it is better to break the API so that we are as close to a vanilla `box` as possible.

**Comment by ericpre** [comment]
@francisco-dlp, would it be possible to have the search functionalities if we were to use `python-box` in hyperspy?

**Comment by francisco-dlp** [comment]
In principle, I don't see why it shouldn't be possible to implement the search features on top of `box`. If we agree that it is a good idea to replace `DictionaryTreeBrowser` with `box`, I can give it a try.

**Comment by ericpre** [comment]
In principle, this is a good idea but in practise, the `DictionaryTreeBrowser` works fine and has worked fine for a long time and is very stable.

It seems that `python-box` breaks API quite often, for example, `python-box` is currently pinned in `rosettasciio` because of that, so we may end up with more maintenance burden, particularly if we are subclassing a class of `python-box`. Ideally, the feature we need should be implemented in `python-box` but this will take time.

I suspect that it could easily turn into a rabbit hole (thinking for example of the "slugify" function to normalise the keys) for little reward. Also would it make to do that when we revisit the metadata structure?

**Comment by francisco-dlp** [comment]
After a quick test  (https://github.com/francisco-dlp/hyperspy/tree/BoxDTB), I agree with @ericpre: the task is less straightforward than I thought and I see how it can easily turn into a rabbit hole. For example, `box.update()` doesn't behave exactly as our `DictionaryTreeBrowser.add_dictionary()`. There may be more subtle differences, so it is definetily not something to implement in a rush.

**Comment by CSSFrancis** [comment]
So do we want to make a Release Candidate for hyperspy 2.0.0 :)


Edit: we should maybe wait for #3255

**Comment by ericpre** [comment]
> Edit: we should maybe wait for https://github.com/hyperspy/hyperspy/pull/3255

Yes, as it is almost there, I will to try to finish it today.

**Comment by CSSFrancis** [comment]
> > Edit: we should maybe wait for #3255
> 
> Yes, as it is almost there, I will to try to finish it today.

No worries :) Thanks for doing that

**Comment by ericpre** [comment]
And possibly #3258 too...

**Comment by ericpre** [comment]
We will need to make a RosettaSciIO release before making the rc of hyperspy to support the split.

Left to be done:
- [x] https://github.com/hyperspy/rosettasciio/pull/180
- [x] https://github.com/hyperspy/hyperspy/pull/3255
- [x] https://github.com/hyperspy/hyperspy/pull/3258

**Comment by francisco-dlp** [comment]
`h5py` runs fine in pyodide and pypy, so I have added it back to the default HyperSpy installation in https://github.com/hyperspy/hyperspy/pull/3260.

**Comment by CSSFrancis** [comment]
Once #3262 is merged then we should be able to make a rc?

**Comment by ericpre** [comment]
Next steps:

1. release `holospy` 0.1: this is the most simple and it should be ready to go. Use it to review standardisation/simplification of packaging and workflow: https://github.com/hyperspy/holospy/pull/14
3. release  2.0rc0 `GUIs` of libraries? I am not sure if a rc is necessary.
5. Create `exspy-demos` repository?
6. release `exspy` 0.1
7. release `hyperspy` 2.0
8. update hyperspy-feedstock
9. package `holospy` and `exspy` on conda-forge

When specifying dependency, we can use `hyperspy>=2.0rc0` and `pip install` will pick up this version without having to specify the `--pre` flag.

In parallel, we should do the hyperspy 1.7.6 release, maybe we should leave it a couple of days in case people want to try swapping between 2.0rc0 and the 1.7.5.dev0 branch? Aim for release on the 17th Nov.?

**Comment by CSSFrancis** [comment]
We should probably think about what it would take to get the subpackages compliant 

- [ ]  pyxem  https://github.com/pyxem/pyxem/pull/969
- [ ] atomap see: https://gitlab.com/atomap/atomap/-/merge_requests/74
    - [ ] Resolve #3265
- [ ] lumispy --> No failing tests with 2.0.0
- [ ] kikchuipy https://github.com/pyxem/kikuchipy/issues/650 
- [ ] particlespy

**Comment by magnunor** [comment]
I'll try to get an Atomap release out this weekend, pinned to HyperSpy 1.x.x.

Then get another release out which works with the HyperSpy 2.0.0 release candidate.

@CSSFrancis: sorry about the slow response, busy times with end of teaching semester + upcoming application deadline in December :)

**Comment by CSSFrancis** [comment]
> @CSSFrancis: sorry about the slow response, busy times with end of teaching semester + upcoming application deadline in December :)

No problem :)  If figured that this is not the most relaxing time with teaching, I've just been pinging you for things that might be of interest.

@ericpre Do we want to push back the 2.0.0 release maybe one more week until we can figure out some of the problems downstream.

**Comment by ericpre** [comment]
I would suggest to make the hyperspy 2.0 release by Wednesday. There is only https://github.com/hyperspy/hyperspy/pull/3281 left! :)

**Comment by ericpre** [comment]
This can be closed now! 😃
